node/core/pvf/src/worker_common.rs

```diff
@@ -199,71 +201,45 @@ where
-	mut stream: UnixStream,
 	cpu_time_start: ProcessTime,
 	timeout: Duration,
-	lock: Arc<AtomicBool>,
+	finished_flag: Arc<AtomicBool>,
 ) {
 	loop {
+		if finished_flag.load(Ordering::Relaxed) {
+			return
+		}
+
 		let cpu_time_elapsed = cpu_time_start.elapsed();
 
 		// Treat the timeout as CPU time, which is less subject to variance due to load.
-		if cpu_time_elapsed > timeout {
-			let result = lock.compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed);
-			if result.is_err() {
-				// Hit the job-completed case first, return from this thread.
-				return
-			}
-
-			// Log if we exceed the timeout.
-			gum::warn!(
-				target: LOG_TARGET,
-				worker_pid = %std::process::id(),
-				"{job_kind} job took {}ms cpu time, exceeded {job_kind} timeout {}ms",
-				cpu_time_elapsed.as_millis(),
-				timeout.as_millis(),
-			);
-
-			// Send back a `TimedOut` error.
-			//
-			// NOTE: This will cause the worker, whether preparation or execution, to be killed by
-			// the host. We do not kill the process here because it would interfere with the proper
-			// handling of this error.
-			let encoded_result = match job_kind {
-				JobKind::Prepare => {
-					let result: Result<(), PrepareError> = Err(PrepareError::TimedOut);
-					result.encode()
-				},
-				JobKind::Execute => {
-					let result = ExecuteResponse::TimedOut;
-					result.encode()
-				},
-			};
-			// If we error here there is nothing we can do apart from log it. The receiving side
-			// will just have to time out.
-			if let Err(err) = framed_send(&mut stream, encoded_result.as_slice()).await {
-				gum::warn!(
-					target: LOG_TARGET,
-					worker_pid = %std::process::id(),
-					"{job_kind} worker -> pvf host: error sending result over the socket: {:?}",
-					err
-				);
-			}
-
-			return
+		if cpu_time_elapsed <= timeout {
+			// Sleep for the remaining CPU time, plus a bit to account for overhead. Note that the sleep
+			// is wall clock time. The CPU clock may be slower than the wall clock.
+			let sleep_interval = timeout.saturating_sub(cpu_time_elapsed) + JOB_TIMEOUT_OVERHEAD;
+			std::thread::sleep(sleep_interval);
+			continue
 		}
 
-		// Sleep for the remaining CPU time, plus a bit to account for overhead. Note that the sleep
-		// is wall clock time. The CPU clock may be slower than the wall clock.
-		let sleep_interval = timeout - cpu_time_elapsed + JOB_TIMEOUT_OVERHEAD;
-		std::thread::sleep(sleep_interval);
+		// Log if we exceed the timeout.
+		gum::warn!(
+			target: LOG_TARGET,
+			worker_pid = %std::process::id(),
+			"{job_kind} job took {}ms cpu time, exceeded {job_kind} timeout {}ms",
+			cpu_time_elapsed.as_millis(),
+			timeout.as_millis(),
+		);
+
+		// Return, signaling that we should send a `TimedOut` error to the host.
+		//
+		// NOTE: This will cause the worker, whether preparation or execution, to be killed by the
+		// host. We do not kill the process here because it would interfere with the proper handling
+		// of this error.
```

Comment:
This should be at the top of the function as docs. So, that someone can see it without reading the code.